### ISSUE TITLE
potential support for tree training

### ISSUE THREAD

**Issue by marsupialtail** [microsoft/hummingbird#522]
I am wondering if there is plan to convert the training phase of these algorithms to matrix operations? Feel free to close this issue if it's not relevant.

**Comment by interesaaat** [comment]
Hey! Yes we can do training with the GEMM strategy. Check [this branch](https://github.com/microsoft/hummingbird/tree/mainterl/fine-tune-trees) if you want to try training.

**Comment by marsupialtail** [comment]
Sorry can you please give me more information? Like what code do I run once I check that branch?

**Comment by interesaaat** [comment]
Sure! There is a test [here](https://github.com/microsoft/hummingbird/blob/mainterl/fine-tune-trees/tests/test_sklearn_gbdt_converter_fine_tune.py) showing how to do training \ fine tuning of a GBDT model. If you are interested in more details, [here](https://arxiv.org/abs/1906.03822) is a paper explaining the technique.

**Comment by marsupialtail** [comment]
Can you give some ballpark performance estimates of using this approach versus say sklearn's tree.fit()? Would I ever want to use this instead of tree.fit()? I would be interested in potentially contributing some code if this area looks promising.

**Comment by interesaaat** [comment]
In general this is slower, if you are looking for performance, even when you are using a GPU. There might be some trick there to boost performance, but we haven't look into this yet (if you want, this could be a good thing to look at). What we found is that this method can return better accuracy.

**Comment by marsupialtail** [comment]
I see that this is code for finetuning. Is training from scratch a worthwhile thing to do? Does the finetuning give you better accuracy?

**Comment by interesaaat** [comment]
In the paper I linked above we weren't able to find cases where training from scratch was better than finetuning. Our intuition is that the trained model already provide good initial parameters for the NN model.
